From 86102deaee3069bf81eb0d25dda9868d9267053f Mon Sep 17 00:00:00 2001
From: Yipeng Zou <zouyipeng@huawei.com>
Date: Thu, 9 Feb 2023 06:58:07 +0000
Subject: [PATCH 901/905] [Huawei] genirq: add printk safe in irq context

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6BO2R

--------------------------------

The printk has the risk of deadlocks in the context of an interrupt.

Introduce printk safe to avoid it.

Fixes: 0112878fff82 ("[Huawei] genirq: Introduce warn log when irq be reentrant")
Signed-off-by: Yipeng Zou <zouyipeng@huawei.com>
---
 kernel/irq/chip.c | 16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/kernel/irq/chip.c b/kernel/irq/chip.c
index 4becd299a929..3fafebfd28a5 100644
--- a/kernel/irq/chip.c
+++ b/kernel/irq/chip.c
@@ -514,9 +514,19 @@ static bool irq_may_run(struct irq_desc *desc)
 		return true;
 
 #ifdef CONFIG_IRQ_REENTRY_WARN
-	if (irqd_get(&desc->irq_data) & IRQD_IRQ_INPROGRESS)
-		pr_warn_ratelimited("irq %u(%lu) may be reentrant in multiple cpus.\n",
-				    desc->irq_data.irq, desc->irq_data.hwirq);
+	if (irqd_get(&desc->irq_data) & IRQD_IRQ_INPROGRESS) {
+		const char *name = NULL;
+
+		if (desc->name)
+			name = desc->name;
+		else if (desc->action)
+			name = desc->action->name;
+
+		printk_safe_enter();
+		pr_warn("irq %u(%s) may be reentrant in multiple cpus.\n",
+			desc->irq_data.irq, name == NULL ? "NULL" : name);
+		printk_safe_exit();
+	}
 #endif
 
 	/*
-- 
2.17.1

